File: src/backend/src/main/java/com/prcins/umbrella/service/user/UserService.java
package com.prcins.umbrella.service.user;

import com.prcins.umbrella.web.dto.UserDTO;
import java.util.Optional;
import java.util.Set;
import java.time.LocalDateTime;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.access.prepost.PreAuthorize;

/**
 * Service interface for user management operations with enhanced security features,
 * Virtual Thread support, and container-aware capabilities.
 * Leverages Spring Boot 3.2.x and Jakarta EE standards for modern enterprise applications.
 */
public interface UserService {

    /**
     * Creates a new user with enhanced security validation.
     * Executes in a Virtual Thread for improved performance under high load.
     *
     * @param userDTO The user data transfer object containing user information
     * @return The created user DTO with security metadata
     * @throws jakarta.validation.ValidationException if validation fails
     * @throws jakarta.persistence.EntityExistsException if username/email already exists
     */
    @PreAuthorize("hasPermission('USER_MANAGEMENT', 'FULL')")
    UserDTO createUser(UserDTO userDTO);

    /**
     * Updates existing user information with optimistic locking.
     * Supports container-aware transaction management.
     *
     * @param userId The ID of the user to update
     * @param userDTO The updated user information
     * @return The updated user DTO with refreshed security context
     * @throws jakarta.persistence.EntityNotFoundException if user not found
     * @throws jakarta.persistence.OptimisticLockException if concurrent modification detected
     */
    @PreAuthorize("hasPermission('USER_MANAGEMENT', 'READ_WRITE')")
    UserDTO updateUser(Long userId, UserDTO userDTO);

    /**
     * Retrieves a user by their unique identifier.
     * Leverages Virtual Thread for non-blocking database operations.
     *
     * @param userId The ID of the user to retrieve
     * @return Optional containing the user if found
     */
    @PreAuthorize("hasPermission('USER_MANAGEMENT', 'READ')")
    Optional<UserDTO> getUserById(Long userId);

    /**
     * Retrieves a paginated list of users with Virtual Thread support.
     * Optimized for container deployment with connection pool awareness.
     *
     * @param pageable Pagination information
     * @return Page of user DTOs
     */
    @PreAuthorize("hasPermission('USER_MANAGEMENT', 'READ')")
    Page<UserDTO> getUsers(Pageable pageable);

    /**
     * Changes user password with enhanced security validation.
     * Implements rate limiting and password history checks.
     *
     * @param userId The ID of the user
     * @param currentPassword The current password
     * @param newPassword The new password
     * @throws jakarta.security.auth.AuthenticationException if current password is invalid
     * @throws jakarta.validation.ValidationException if password requirements not met
     */
    @PreAuthorize("hasPermission('USER_MANAGEMENT', 'READ_WRITE') or #userId == authentication.principal.id")
    void changePassword(Long userId, String currentPassword, String newPassword);

    /**
     * Updates user roles with atomic operations.
     * Supports container-aware security context propagation.
     *
     * @param userId The ID of the user
     * @param roleIds Set of role IDs to assign
     * @return Updated user DTO with new role assignments
     * @throws jakarta.persistence.EntityNotFoundException if user or roles not found
     */
    @PreAuthorize("hasPermission('USER_MANAGEMENT', 'FULL')")
    UserDTO updateUserRoles(Long userId, Set<Long> roleIds);

    /**
     * Deactivates a user account.
     * Maintains audit trail for compliance requirements.
     *
     * @param userId The ID of the user to deactivate
     * @throws jakarta.persistence.EntityNotFoundException if user not found
     */
    @PreAuthorize("hasPermission('USER_MANAGEMENT', 'FULL')")
    void deactivateUser(Long userId);

    /**
     * Records user login attempt with security metadata.
     * Supports distributed session management in containerized environments.
     *
     * @param userId The ID of the user
     * @param loginTimestamp The timestamp of the login attempt
     * @param successful Whether the login attempt was successful
     */
    void recordLoginAttempt(Long userId, LocalDateTime loginTimestamp, boolean successful);

    /**
     * Validates user credentials with rate limiting.
     * Optimized for Virtual Thread execution.
     *
     * @param username The username to validate
     * @param password The password to validate
     * @return true if credentials are valid
     */
    boolean validateCredentials(String username, String password);

    /**
     * Checks if a username is available.
     * Executes in Virtual Thread for improved responsiveness.
     *
     * @param username The username to check
     * @return true if username is available
     */
    boolean isUsernameAvailable(String username);

    /**
     * Checks if an email address is available.
     * Supports container-aware caching.
     *
     * @param email The email to check
     * @return true if email is available
     */
    boolean isEmailAvailable(String email);
}

File: src/web/src/services/user.service.ts
/**
 * User Management Service
 * Provides user management operations with Spring Boot 3.2.x security integration
 * and Virtual Thread support for enhanced performance.
 * @version 1.0.0
 */

import { ApiService } from './api.service';
import { 
  User, 
  Role, 
  Permission, 
  AccessLevel, 
  ResourceType,
  UserApiResponse,
  AuthResponse,
  AuthRequest 
} from '../types/user.types';
import { PaginationParams, ApiResponse, PaginatedResponse } from '../types/common.types';
import { USER_ENDPOINTS, AUTH_ENDPOINTS } from '../constants/api.constants';

/**
 * Interface for creating new users with Spring Security validation
 */
interface CreateUserDto {
  username: string;
  email: string;
  firstName: string;
  lastName: string;
  password: string;
  roles: number[];
}

/**
 * Interface for updating user details
 */
interface UpdateUserDto {
  email?: string;
  firstName?: string;
  lastName?: string;
  roles?: number[];
  enabled?: boolean;
}

/**
 * Service class for user management with Spring Security integration
 * and Virtual Thread support for enhanced performance
 */
export class UserService {
  private readonly baseUrl: string = '/users';

  constructor(private readonly apiService: ApiService) {}

  /**
   * Retrieves paginated list of users with Spring Security filtering
   * @param params Pagination and filtering parameters
   * @returns Promise with paginated user list
   */
  public async getUsers(params: PaginationParams): Promise<ApiResponse<PaginatedResponse<User>>> {
    const queryParams = new URLSearchParams({
      page: params.page.toString(),
      size: params.size.toString(),
      sort: params.sort.map(s => `${s.field},${s.direction}`).join(',')
    });

    if (params.filter) {
      Object.entries(params.filter).forEach(([key, value]) => {
        queryParams.append(`filter[${key}]`, String(value));
      });
    }

    return this.apiService.get<PaginatedResponse<User>>(
      `${this.baseUrl}?${queryParams.toString()}`
    );
  }

  /**
   * Creates new user with Spring Security validation
   * @param userData User creation data
   * @returns Promise with created user details
   */
  public async createUser(userData: CreateUserDto): Promise<UserApiResponse> {
    return this.apiService.post<User>(this.baseUrl, userData);
  }

  /**
   * Updates existing user with role validation
   * @param userId User ID
   * @param userData Update data
   * @returns Promise with updated user details
   */
  public async updateUser(userId: number, userData: UpdateUserDto): Promise<UserApiResponse> {
    return this.apiService.put<User>(`${this.baseUrl}/${userId}`, userData);
  }

  /**
   * Deletes user with permission check
   * @param userId User ID
   * @returns Promise with deletion status
   */
  public async deleteUser(userId: number): Promise<ApiResponse<void>> {
    return this.apiService.delete(`${this.baseUrl}/${userId}`);
  }

  /**
   * Authenticates user with Spring Security
   * @param credentials Login credentials
   * @returns Promise with authentication response
   */
  public async login(credentials: AuthRequest): Promise<ApiResponse<AuthResponse>> {
    return this.apiService.post<AuthResponse>(AUTH_ENDPOINTS.LOGIN, credentials);
  }

  /**
   * Logs out current user and invalidates session
   * @returns Promise with logout status
   */
  public async logout(): Promise<ApiResponse<void>> {
    return this.apiService.post(AUTH_ENDPOINTS.LOGOUT);
  }

  /**
   * Retrieves current user profile
   * @returns Promise with user profile data
   */
  public async getCurrentUser(): Promise<UserApiResponse> {
    return this.apiService.get<User>(USER_ENDPOINTS.GET_PROFILE);
  }

  /**
   * Updates current user's password with security validation
   * @param currentPassword Current password
   * @param newPassword New password
   * @returns Promise with update status
   */
  public async changePassword(
    currentPassword: string,
    newPassword: string
  ): Promise<ApiResponse<void>> {
    return this.apiService.post(USER_ENDPOINTS.CHANGE_PASSWORD, {
      currentPassword,
      newPassword
    });
  }

  /**
   * Checks if user has specific permission using Spring Security's @PreAuthorize
   * @param userId User ID
   * @param permission Permission to check
   * @param resourceType Resource type for permission
   * @returns Promise with permission check result
   */
  public async hasPermission(
    userId: number,
    permission: Permission,
    resourceType: ResourceType
  ): Promise<boolean> {
    try {
      const response = await this.apiService.get<{ hasPermission: boolean }>(
        `${this.baseUrl}/${userId}/permissions/check`,
        {
          params: {
            permission: permission.name,
            resourceType
          }
        }
      );
      return response.data.hasPermission;
    } catch (error) {
      console.error('Permission check failed:', error);
      return false;
    }
  }

  /**
   * Assigns roles to user with validation
   * @param userId User ID
   * @param roleIds Role IDs to assign
   * @returns Promise with updated user details
   */
  public async assignRoles(userId: number, roleIds: number[]): Promise<UserApiResponse> {
    return this.apiService.post<User>(`${this.baseUrl}/${userId}/roles`, { roleIds });
  }

  /**
   * Retrieves user's effective permissions across all roles
   * @param userId User ID
   * @returns Promise with list of effective permissions
   */
  public async getEffectivePermissions(userId: number): Promise<ApiResponse<Permission[]>> {
    return this.apiService.get<Permission[]>(`${this.baseUrl}/${userId}/permissions`);
  }

  /**
   * Validates user's access level for specific resource
   * @param userId User ID
   * @param resourceType Resource type
   * @returns Promise with access level
   */
  public async getResourceAccessLevel(
    userId: number,
    resourceType: ResourceType
  ): Promise<ApiResponse<AccessLevel>> {
    return this.apiService.get<AccessLevel>(
      `${this.baseUrl}/${userId}/access-level/${resourceType}`
    );
  }
}